src/main/java/hudson/plugins/gradle/injection/MavenOptsSetter.java

```diff
@@ -17,54 +17,81 @@ class MavenOptsSetter {
 
     private final Set<String> keys;
+    private final Set<String> requiredKeys;
 
-    public MavenOptsSetter(String... keys) {
-        this.keys = ImmutableSet.copyOf(keys);
+    public MavenOptsSetter(SystemProperty.Key... keys) {
+        this.keys =
+            Arrays.stream(keys)
+                .map(k -> k.name)
+                .collect(
+                    Collectors.collectingAndThen(Collectors.toSet(), Collections::unmodifiableSet));
+        this.requiredKeys =
+            Arrays.stream(keys)
+                .filter(k -> k.required)
+                .map(k -> k.name)
+                .collect(
+                    Collectors.collectingAndThen(Collectors.toSet(), Collections::unmodifiableSet));
     }
 
-    void appendIfMissing(Node node, List<SystemProperty> systemProperties) throws IOException, InterruptedException {
-        String geMavenOpts =
+    void append(Node node, List<SystemProperty> systemProperties) throws IOException, InterruptedException {
+        String additionalProperties =
             systemProperties
                 .stream()
                 .map(SystemProperty::asString)
                 .collect(Collectors.joining(SPACE));
 
-        String mavenOpts = removeSystemProperties(getMavenOpts(node)) + SPACE + geMavenOpts;
+        String mavenOpts =
+            filtered(getCurrentMavenOpts(node))
+                .map(current -> String.join(SPACE, current, additionalProperties))
+                .orElse(additionalProperties);
+
         EnvUtil.setEnvVar(node, MAVEN_OPTS_VAR, mavenOpts);
     }
 
-    void remove(Node node) throws IOException, InterruptedException {
-        String mavenOpts = removeSystemProperties(getMavenOpts(node));
+    void removeIfNeeded(Node node) throws IOException, InterruptedException {
+        String currentMavenOpts = getCurrentMavenOpts(node);
+        if (currentMavenOpts == null || currentMavenOpts.isEmpty()) {
+            return;
+        }
+
+        boolean hasAllRequiredKeys = requiredKeys.stream().allMatch(currentMavenOpts::contains);
+        if (!hasAllRequiredKeys) {
+            return;
+        }
+
+        String mavenOpts = filtered(currentMavenOpts).orElse(null);
         EnvUtil.setEnvVar(node, MAVEN_OPTS_VAR, mavenOpts);
     }
 
-    private String getMavenOpts(Node node) throws IOException, InterruptedException {
-        EnvVars nodeEnvVars = EnvVars.getRemote(node.getChannel());
-        return nodeEnvVars.get(MAVEN_OPTS_VAR);
+    @Nullable
+    private static String getCurrentMavenOpts(Node node) {
+        return EnvUtil.getEnv(node, MAVEN_OPTS_VAR);
     }
 
-    private String removeSystemProperties(String mavenOpts) throws RuntimeException {
+    private Optional<String> filtered(@Nullable String mavenOpts) throws RuntimeException {
         return Optional.ofNullable(mavenOpts)
-            .map(this::filterMavenOpts)
-            .orElse("");
+            .map(this::filterMavenOpts);
```

Comment:
Same, how about changing the return result of `filterMavenOpts` by `Optional<String>` and `flatMap` here ?